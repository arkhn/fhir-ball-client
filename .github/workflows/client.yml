name: pyrog-client

on: [push]

jobs:
  build:
    name: Build Pyrog client
    runs-on: ubuntu-18.04

    strategy:
      matrix:
        node-version: [12.x]

    steps:
      - uses: actions/checkout@v1

      - name: Use Node.js ${{ matrix.node-version }}
        uses: actions/setup-node@v1
        with:
          node-version: ${{ matrix.node-version }}

      - name: Cache node modules
        uses: actions/cache@v1
        with:
          path: client/node_modules
          key: ${{ runner.OS }}-${{ matrix.node-version}}-node-modules-${{ hashFiles('client/yarn.lock') }}

      - name: Install Dependencies
        run: |
          cd client
          yarn install

      - name: Build
        run: |
          cd client
          yarn build

  # Same cleaning need to be done, see Issue #105
  #      - name: Lint
  #        run: |
  #          cd client
  #          yarn lint

  publish:
    name: Build and publish
    env:
      GITHUB_REGISTRY: docker.pkg.github.com
      IMAGE: arkhn/pyrog-client
    runs-on: ubuntu-18.04
    needs: build
    if: github.ref == 'refs/heads/master'
    steps:
      - uses: actions/checkout@v1

      - name: Build
        run: |
          cd client/
          yarn docker:build:staging

      - name: Tag docker images
        run: |
          docker tag $IMAGE $IMAGE:$GITHUB_SHA
          docker tag $IMAGE $GITHUB_REGISTRY/${{ github.repository }}/pyrog-client:latest
          docker tag $IMAGE $GITHUB_REGISTRY/${{ github.repository }}/pyrog-client:$GITHUB_SHA

      - name: Publish to the docker hub
        run: |
          echo "${{ secrets.DOCKER_PASSWORD }}" |  docker login -u ${{ secrets.DOCKER_LOGIN }} --password-stdin
          docker push $IMAGE:latest
          docker push $IMAGE:$GITHUB_SHA

      - name: Publish to Github Packages
        run: |
          echo "${{ secrets.GITHUB_TOKEN }}" | docker login $GITHUB_REGISTRY -u ${{ github.actor }} --password-stdin
          docker push $GITHUB_REGISTRY/${{ github.repository }}/pyrog-client:latest
          docker push $GITHUB_REGISTRY/${{ github.repository }}/pyrog-client:$GITHUB_SHA
