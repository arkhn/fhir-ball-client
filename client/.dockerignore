# ignore everything
**/*

# except
!.dockerignore
!Dockerfile
!yarn.lock
!package.json
!public
!src
!tsconfig.json
!tsconfig.build.json
!.env*
!build