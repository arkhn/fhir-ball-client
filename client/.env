PORT=1729
REACT_APP_HTTP_BACKEND_URL=http://localhost:1000
REACT_APP_RIVER_URL=https://staging.arkhn.com/river
REACT_APP_PAGAI_URL=https://staging.arkhn.com/pagai
REACT_APP_CLEANING_SCRIPTS_URL=https://staging.arkhn.com/scripts
REACT_APP_FHIR_API_URL=https://staging.arkhn.com/api
