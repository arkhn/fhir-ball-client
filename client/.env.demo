PORT=1729
REACT_APP_HTTP_BACKEND_URL=https://fhir.demo.arkhn.com/pyrog
REACT_APP_FHIRPIPE_URL=https://fhir.demo.arkhn.com/pipe
REACT_APP_CLEANING_SCRIPTS_URL=https://fhir.demo.arkhn.com/scripts
REACT_APP_FHIR_API_URL=https://fhir.demo.arkhn.com/api
REACT_APP_PAGAI_URL=https://fhir.demo.arkhn.com/pagai
