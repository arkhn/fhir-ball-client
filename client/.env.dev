PORT=1729
REACT_APP_HTTP_BACKEND_URL=https://dev.arkhn.com/pyrog-api
REACT_APP_RIVER_URL=https://dev.arkhn.com/river
REACT_APP_CLEANING_SCRIPTS_URL=https://dev.arkhn.com/scripts
REACT_APP_FHIR_API_URL=https://dev.arkhn.com/api
REACT_APP_PAGAI_URL=https://dev.arkhn.com/pagai
REACT_APP_CLIENT_ID=local-pyrog-client
REACT_APP_CLIENT_SECRET=pyrog-client
REACT_APP_AUTH_URL=https://dev.arkhn.com/hydra/oauth2/auth
REACT_APP_TOKEN_URL=https://dev.arkhn.com/hydra/oauth2/token
REACT_APP_REVOKE_URL=https://dev.arkhn.com/hydra/oauth2/revoke
REACT_APP_LOGOUT_URL=https://dev.arkhn.com/hydra/oauth2/sessions/logout
REACT_APP_LOGIN_REDIRECT_URL=https://dev.arkhn.com/pyrog/
REACT_APP_LOGOUT_REDIRECT_URL=https://dev.arkhn.com/pyrog/
