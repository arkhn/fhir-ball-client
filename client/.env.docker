#pyrog-client
PORT=1729
REACT_APP_HTTP_BACKEND_URL=http://pyrog-server:1000
REACT_APP_FHIRPIPE_URL=http://localhost:3000
REACT_APP_CLEANING_SCRIPTS_URL=http://cleaning-scripts:5000
REACT_APP_PAGAI_URL=http://pagai:4000

#cleaning-scripts
GITHUB_TOKEN=<token>


#pagai
PYROG_URL=http://pyrog-server:4000
PYROG_TOKEN=<token>
