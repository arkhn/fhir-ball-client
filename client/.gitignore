node_modules/
dist/*
build/

coverage
yarn-error.log

.DS_Store