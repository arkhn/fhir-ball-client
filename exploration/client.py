import os
import psycopg2
import yaml

configFileName = './conf.custom.yml' if os.path.exists('./conf.custom.yml') else './conf.default.yml'

def main():
    """Initiates connection with distant postgres database."""

    with open(configFileName) as config_file:
        try:
            # Load config
            config = yaml.safe_load(config_file)
            connection = None

            try:
                # Initialise client
                connection = psycopg2.connect(
                    host=config['db_host'],
                    port=config['db_port'],
                    database=config['db_name'],
                    user=config['db_user'],
                    password=config['db_password']
                )

                # Run SQL query
                cursor = connection.cursor()
                cursor.execute("SELECT * from patients limit 10;")
                data = cursor.fetchall()
                print(data)

            except (Exception, psycopg2.Error) as error :
                print ("Error while connecting to PostgreSQL", error)
            finally:
                # Closing database connection.
                if(connection):
                    cursor.close()
                    connection.close()
                    print("PostgreSQL connection is closed")

        except yaml.YAMLError as error:
            print(error)

if __name__ == "__main__":
    main()