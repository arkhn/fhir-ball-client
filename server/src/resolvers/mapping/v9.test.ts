import { PrismaClient } from '@prisma/client'

import importMappingV9 from './v9'
import * as mappingV9 from '../../../test/fixtures/chimio-mapping-v9.json'

const mockCreateResource = jest.fn()

const mockFindManyUser = jest.fn(() => [{ email: 'admin@arkhn.com' }])
jest.mock('@prisma/client', () => ({
  PrismaClient: jest.fn().mockImplementation(() => ({
    resource: {
      create: (data: any) => mockCreateResource(data),
    },
    user: {
      findMany: mockFindManyUser,
    },
  })),
}))

describe('import mapping V9', () => {
  const sourceId = '01234567'
  const { resources } = mappingV9 as any

  beforeEach(() => {
    mockCreateResource.mockClear()
  })

  it('should send a query per resource', async () => {
    await importMappingV9(new PrismaClient(), sourceId, resources)
    expect(mockCreateResource).toHaveBeenCalledTimes(resources.length)
    expect(mockCreateResource.mock.calls[0]).toMatchSnapshot() // EpisodeOfCare - HopitalStay
    expect(mockCreateResource.mock.calls[1]).toMatchSnapshot() // Patient
  })

  it('should raise an error if importing mapping with unexisting comment author', async () => {
    mockFindManyUser.mockReturnValueOnce([{ email: 'user@arkhn.com' }])
    const t = importMappingV9(new PrismaClient(), sourceId, resources)
    expect(t).rejects.toThrowError(
      'trying to import a mapping with unexisting comment author',
    )
  })

  it('should have cleaned the resource and attributes', async () => {
    await importMappingV9(new PrismaClient(), sourceId, resources)
    expect(mockCreateResource.mock.calls[0][0]).toEqual({
      data: {
        label: resources[0].label,
        primaryKeyTable: resources[0].primaryKeyTable,
        primaryKeyColumn: resources[0].primaryKeyColumn,
        source: {
          connect: {
            id: sourceId,
          },
        },
        definitionId: 'EpisodeOfCare',
        attributes: {
          create: expect.arrayContaining([
            {
              path: 'period.start',
              definitionId: 'dateTime',
              sliceName: null,
              inputGroups: {
                create: expect.arrayContaining([
                  {
                    mergingScript: 'merge_concat',
                    inputs: {
                      create: expect.any(Array),
                    },
                    conditions: {
                      create: expect.arrayContaining([
                        {
                          action: 'EXCLUDE',
                          value: '-1',
                          sqlValue: {
                            create: { table: 'SEJOUR', column: 'DATEIN' },
                          },
                        },
                      ]),
                    },
                  },
                ]),
              },
            },
            {
              path: 'managingOrganization.reference',
              comments: {
                create: [
                  {
                    author: {
                      connect: {
                        email: 'admin@arkhn.com',
                      },
                    },
                    content: 'test',
                    createdAt: '2020-04-02T08:53:07.997Z',
                  },
                ],
              },
              sliceName: null,
              definitionId: 'string',
              inputGroups: {
                create: expect.arrayContaining([
                  {
                    mergingScript: null,
                    inputs: {
                      create: expect.any(Array),
                    },
                  },
                ]),
              },
            },
          ]),
        },
        filters: {
          create: [
            {
              value: '200',
              relation: '<=',
              sqlColumn: {
                create: {
                  column: 'CHAMBRE',
                  table: 'SEJOUR',
                },
              },
            },
          ],
        },
      },
    })
  })
})
