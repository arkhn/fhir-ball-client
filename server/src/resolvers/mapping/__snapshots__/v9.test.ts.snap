// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`import mapping V9 should send a query per resource 1`] = `
Array [
  Object {
    "data": Object {
      "attributes": Object {
        "create": Array [
          Object {
            "definitionId": "string",
            "inputGroups": Object {
              "create": Array [
                Object {
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "sqlValue": Object {
                          "create": Object {
                            "column": "LIBELTW",
                            "joins": Object {
                              "create": Array [
                                Object {
                                  "tables": Object {
                                    "create": Array [
                                      Object {
                                        "column": "TWCLEUNIK",
                                        "table": "SEJOUR",
                                      },
                                      Object {
                                        "column": "TWCLEUNIK",
                                        "table": "TYPESEJ",
                                      },
                                    ],
                                  },
                                },
                              ],
                            },
                            "table": "TYPESEJ",
                          },
                        },
                        "staticValue": null,
                      },
                    ],
                  },
                  "mergingScript": null,
                },
              ],
            },
            "path": "type[0].coding[0].display",
            "sliceName": null,
          },
          Object {
            "definitionId": "Reference",
            "path": "patient",
            "sliceName": null,
          },
          Object {
            "definitionId": "uri",
            "inputGroups": Object {
              "create": Array [
                Object {
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "staticValue": "http://terminology.arkhn.com/ck8iiui3g0873euutm08fnyrf",
                      },
                    ],
                  },
                  "mergingScript": null,
                },
              ],
            },
            "path": "patient.identifier.system",
            "sliceName": null,
          },
          Object {
            "definitionId": "string",
            "inputGroups": Object {
              "create": Array [
                Object {
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "sqlValue": Object {
                          "create": Object {
                            "column": "PACLEUNIK",
                            "table": "SEJOUR",
                          },
                        },
                        "staticValue": null,
                      },
                    ],
                  },
                  "mergingScript": null,
                },
              ],
            },
            "path": "patient.identifier.value",
            "sliceName": null,
          },
          Object {
            "definitionId": "Identifier",
            "path": "managingOrganization.identifier",
            "sliceName": null,
          },
          Object {
            "definitionId": "string",
            "inputGroups": Object {
              "create": Array [
                Object {
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "sqlValue": Object {
                          "create": Object {
                            "column": "SERVICE",
                            "table": "SEJOUR",
                          },
                        },
                        "staticValue": null,
                      },
                    ],
                  },
                  "mergingScript": null,
                },
              ],
            },
            "path": "managingOrganization.identifier.value",
            "sliceName": null,
          },
          Object {
            "definitionId": "dateTime",
            "inputGroups": Object {
              "create": Array [
                Object {
                  "conditions": Object {
                    "create": Array [
                      Object {
                        "action": "EXCLUDE",
                        "relation": "EQ",
                        "sqlValue": Object {
                          "create": Object {
                            "column": "DATEIN",
                            "table": "SEJOUR",
                          },
                        },
                        "value": "-1",
                      },
                    ],
                  },
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": "clean_date",
                        "sqlValue": Object {
                          "create": Object {
                            "column": "DATEIN",
                            "table": "SEJOUR",
                          },
                        },
                        "staticValue": null,
                      },
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "sqlValue": Object {
                          "create": Object {
                            "column": "HEUREIN",
                            "table": "SEJOUR",
                          },
                        },
                        "staticValue": null,
                      },
                    ],
                  },
                  "mergingScript": "merge_concat",
                },
                Object {
                  "conditions": Object {
                    "create": Array [
                      Object {
                        "action": "INCLUDE",
                        "relation": "GT",
                        "sqlValue": Object {
                          "create": Object {
                            "column": "DATEIN",
                            "table": "SEJOUR",
                          },
                        },
                        "value": "-1",
                      },
                    ],
                  },
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "staticValue": "NONE",
                      },
                    ],
                  },
                  "mergingScript": null,
                },
              ],
            },
            "path": "period.start",
            "sliceName": null,
          },
          Object {
            "definitionId": "string",
            "inputGroups": Object {
              "create": Array [
                Object {
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "sqlValue": Object {
                          "create": Object {
                            "column": "SECLEUNIK",
                            "table": "SEJOUR",
                          },
                        },
                        "staticValue": null,
                      },
                    ],
                  },
                  "mergingScript": null,
                },
              ],
            },
            "path": "identifier[0].value",
            "sliceName": null,
          },
          Object {
            "definitionId": "code",
            "inputGroups": Object {
              "create": Array [
                Object {
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "sqlValue": Object {
                          "create": Object {
                            "column": "ETATSEJ",
                            "table": "SEJOUR",
                          },
                        },
                        "staticValue": null,
                      },
                    ],
                  },
                  "mergingScript": null,
                },
              ],
            },
            "path": "status",
            "sliceName": null,
          },
          Object {
            "definitionId": "CodeableConcept",
            "path": "type",
            "sliceName": null,
          },
          Object {
            "definitionId": "string",
            "inputGroups": Object {
              "create": Array [
                Object {
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "staticValue": "internal",
                      },
                    ],
                  },
                  "mergingScript": null,
                },
              ],
            },
            "path": "patient.reference",
            "sliceName": null,
          },
          Object {
            "comments": Object {
              "create": Array [
                Object {
                  "author": Object {
                    "connect": Object {
                      "email": "admin@arkhn.com",
                    },
                  },
                  "content": "test",
                  "createdAt": "2020-04-02T08:53:07.997Z",
                },
              ],
            },
            "definitionId": "string",
            "inputGroups": Object {
              "create": Array [
                Object {
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "staticValue": "internal",
                      },
                    ],
                  },
                  "mergingScript": null,
                },
              ],
            },
            "path": "managingOrganization.reference",
            "sliceName": null,
          },
          Object {
            "definitionId": "Period",
            "path": "period",
            "sliceName": null,
          },
          Object {
            "definitionId": "dateTime",
            "inputGroups": Object {
              "create": Array [
                Object {
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "sqlValue": Object {
                          "create": Object {
                            "column": "HEUREOUT",
                            "table": "SEJOUR",
                          },
                        },
                        "staticValue": null,
                      },
                      Object {
                        "conceptMapId": null,
                        "script": "clean_date",
                        "sqlValue": Object {
                          "create": Object {
                            "column": "DATEOUT",
                            "table": "SEJOUR",
                          },
                        },
                        "staticValue": null,
                      },
                    ],
                  },
                  "mergingScript": "merge_concat",
                },
              ],
            },
            "path": "period.end",
            "sliceName": null,
          },
          Object {
            "definitionId": "Coding",
            "path": "type[0].coding",
            "sliceName": null,
          },
          Object {
            "definitionId": "Identifier",
            "path": "identifier",
            "sliceName": null,
          },
          Object {
            "definitionId": "string",
            "inputGroups": Object {
              "create": Array [
                Object {
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "sqlValue": Object {
                          "create": Object {
                            "column": "ADMISSION",
                            "table": "SEJOUR",
                          },
                        },
                        "staticValue": null,
                      },
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "sqlValue": Object {
                          "create": Object {
                            "column": "ADMISSION2",
                            "table": "SEJOUR",
                          },
                        },
                        "staticValue": null,
                      },
                    ],
                  },
                  "mergingScript": "select_first_not_empty",
                },
              ],
            },
            "path": "identifier[1].value",
            "sliceName": null,
          },
          Object {
            "definitionId": "code",
            "inputGroups": Object {
              "create": Array [
                Object {
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "sqlValue": Object {
                          "create": Object {
                            "column": "TWCLEUNIK",
                            "table": "SEJOUR",
                          },
                        },
                        "staticValue": null,
                      },
                    ],
                  },
                  "mergingScript": null,
                },
              ],
            },
            "path": "type[0].coding[0].code",
            "sliceName": null,
          },
          Object {
            "definitionId": "Identifier",
            "path": "patient.identifier",
            "sliceName": null,
          },
          Object {
            "definitionId": "Reference",
            "path": "managingOrganization",
            "sliceName": null,
          },
          Object {
            "definitionId": "uri",
            "inputGroups": Object {
              "create": Array [
                Object {
                  "inputs": Object {
                    "create": Array [
                      Object {
                        "conceptMapId": null,
                        "script": null,
                        "staticValue": "http://terminology.arkhn.com/ck8iiui3g0872euut6adjvlol",
                      },
                    ],
                  },
                  "mergingScript": null,
                },
              ],
            },
            "path": "managingOrganization.identifier.system",
            "sliceName": null,
          },
        ],
      },
      "definitionId": "EpisodeOfCare",
      "filters": Object {
        "create": Array [
          Object {
            "relation": "<=",
            "sqlColumn": Object {
              "create": Object {
                "column": "CHAMBRE",
                "table": "SEJOUR",
              },
            },
            "value": "200",
          },
        ],
      },
      "label": "Hospital Stay",
      "primaryKeyColumn": "SECLEUNIK",
      "primaryKeyTable": "SEJOUR",
      "source": Object {
        "connect": Object {
          "id": "01234567",
        },
      },
    },
  },
]
`;

exports[`import mapping V9 should send a query per resource 2`] = `
Array [
  Object {
    "data": Object {
      "attributes": Object {
        "create": Array [],
      },
      "definitionId": "Organization",
      "filters": Object {
        "create": Array [],
      },
      "label": "Orga",
      "primaryKeyColumn": "org_id",
      "primaryKeyTable": "org",
      "source": Object {
        "connect": Object {
          "id": "01234567",
        },
      },
    },
  },
]
`;
